Cython/Utility/Optimize.c

```diff
@@ -386,55 +386,72 @@ static CYTHON_INLINE PyObject* __Pyx_dict_iterator(PyObject* iterable, int is_di
     return PyObject_GetIter(iterable);
 }
 
-static CYTHON_INLINE int __Pyx_dict_iter_next(
```

Comment:
The diff looks much much worse than it is. All I did was pick out the contents of the `source_is_dict` block into a separation function, and then call that function inside a critical section....